src/emacs.d/configuration.org
My emacs configuration! Enjoy
* Basic
  #+BEGIN_SRC emacs-lisp

(require 'pallet)
(pallet-mode t)

;; load use-package
(eval-when-compile (require 'use-package))
(require 'diminish)
(require 'bind-key)


  #+END_SRC
* Setup

** SLQ ORACLE
   #+begin_src emacs-lisp
(setq sql-oracle-login-params
      '((user     :default "system")
        (password :default (password-read "Password:" "oracle"))
        (database :default "localhost:49161/xe")))
   #+end_src

** DEBUG
   #+begin_src emacs-lisp
(setq debug-on-error t)
   #+end_src

** HideShow Folding
   #+begin_src emacs-lisp
(use-package hideshow
  :diminish hs-minor-mode
  :config
  (defun hs-hide-stuff ()
    (hs-minor-mode 1)
    (hs-hide-level 3))
  (add-hook 'js2-mode-hook #'hs-hide-stuff )
  (add-hook 'web-mode-hook 'hs-minor-mode )
  (add-hook 'go-mode-hook #'hs-hide-stuff )
  (add-hook 'java-mode-hook #'hs-hide-stuff )
  (add-hook 'emacs-lisp-mode-hook #'hs-hide-stuff )
  (add-hook 'lisp-mode-hook #'hs-hide-stuff )
  (add-hook 'python-mode-hook #'hs-hide-stuff )
  )
   #+end_src

** StrightPush
   #+begin_src emacs-lisp
(defun cesco/automatic-push ()
  (interactive)
  (magit-run-git "commit" "--all"
         (format-time-string "--message=%F %R"))
  (let ((current-branch (magit-get-current-branch)))
    (message "PUSHING!!!!")
    (magit-git-push current-branch
                    (concat "origin/" current-branch)
           nil )
    ))
   #+end_src

** Insert SemiColon
   #+begin_src emacs-lisp

(defun cesco/insert-semicolon ()
  (interactive)
  (evil-append-line "1")
  (insert ";")
  (normal-mode)
(pop-global-mark)
  )
   #+end_src

** Set personal information
   I am a gopher!
   #+BEGIN_SRC emacs-lisp
   (setq user-full-name "Francesco Antonello Ferraro"
        user-mail-address "francescoaferraro@gmail.com "
        calendar-latitude 42.2
        calendar-longitude -71.1
        calendar-location-name "Cambridge, MA")
   #+END_SRC

   #+RESULTS:
   : Cambridge, MA

** Kill other bufferso
   #+BEGIN_SRC emacs-lisp
     (defun cesco/kill-other-buffers ()
       "Kill all other buffers."
       (interactive)
       (mapc 'kill-buffer (delq (current-buffer) (buffer-list))))
   #+END_SRC
** Osx
   #+BEGIN_SRC emacs-lisp
(setq mac-option-modifier 'meta)
(setq mac-command-modifier 'control)
(setq mac-control-modifier 'super)
   #+END_SRC
** REplace region when typing
   #+BEGIN_SRC emacs-lisp
(delete-selection-mode +1)
   #+END_SRC
** PDF

   #+BEGIN_SRC emacs-lisp
(setq doc-view-continuous t)
   #+END_SRC

** Hide Scroll Bar
   #+begin_src emacs-lisp
(scroll-bar-mode -1)
   #+end_src
** Errors

   #+BEGIN_SRC emacs-lisp
(setq help-at-pt-display-when-idle t)
(setq help-at-pt-timer-delay 0.1)
(help-at-pt-set-timer)
   #+END_SRC

** FontSize

   #+BEGIN_SRC emacs-lisp

(cond
 ((string-equal system-type "gnu/linux")
  (set-face-attribute 'default nil :height 115))
 ((string-equal system-type "darwin")
  (set-face-attribute 'default nil :height 150))
 )

   #+END_SRC

** Shift+ Enter

   #+BEGIN_SRC emacs-lisp
     (defun cesco/insert-empty-line ()
       (interactive)
       (newline))
     (global-set-key [(shift return)] 'cesco/insert-empty-line)
   #+END_SRC
** tabbar

   #+BEGIN_SRC emacs-lisp
(use-package tabbar
  :after evil
  :init
  (setq tabbar-cycle-scope 'tabs)
  (setq tabbar-background-color nil)
  (setq tabbar-separator (quote (1.0)))
  (setq tool-bar-mode nil)
  :config
  (setq tabbar-background-color "#393939")
  (custom-set-faces
   '(tabbar-default ((t (:inherit variable-pitch :height 1.4 :background "#393939" :foreground "black" :weight bold))))
   '(tabbar-button ((t (:inherit tabbar-default :foreground "#f2f0ec"))))
   '(tabbar-button-highlight ((t (:inherit tabbar-default))))
   '(tabbar-highlight ((t (:underline t))))
   '(tabbar-selected ((t (:inherit tabbar-default :background "#393939" :foreground "#ffcc66"))))
   '(tabbar-separator ((t (:inherit tabbar-default :background "#393939"))))
   '(tabbar-unselected ((t (:inherit tabbar-default)))))
  (global-set-key (kbd "s-{") 'tabbar-backward-group)
  (global-set-key (kbd "s-}") 'tabbar-forward-group)
  (global-set-key (kbd "s-[") 'tabbar-backward)
  (global-set-key (kbd "s-]") 'tabbar-forward)
  (defun my-tabbar-buffer-groups-by-project ()
    (list
     (cond
      ((memq major-mode '(mu4e-view-mode mu4e-main-mode mu4e-headers-mode mu4e-view-raw-mode
					 twittering-mode weibo-timeline-mode
					 jabber-roster-mode jabber-chat-mode erc-mode douban-music-mode
					 ))
       "Activity"
       )
      ((memq major-mode '(eshell-mode term-mode shell-mode))
       (if (projectile-project-p) (projectile-project-name) "Common")
       )
      ((string-equal "*" (substring (buffer-name) 0 1))
       "Emacs"
       )
      ((memq major-mode '(fundamental-mode))
       "Emacs"
       )
      (t
       (if (projectile-project-p) (projectile-project-name) "Common")
       )
      )))

  (defun my-tabbar-buffer-groups-by-all ()
    (list
     (cond
      ((string-equal "*" (substring (buffer-name) 0 1))
       "Emacs"
       )
      (t "All"))))

  (setq tabbar-buffer-groups-function 'my-tabbar-buffer-groups-by-project)
  (tabbar-mode 1))
   #+END_SRC
** Indent
   #+BEGIN_SRC emacs-lisp

     (defun cesco/indent ()
       (interactive)
       (mark-whole-buffer)
       (indent-region (point-min) (point-max) nil)
       (pop-global-mark)
       )

   #+END_SRC

** PATH

   #+BEGIN_SRC emacs-lisp

(when (memq window-system '(mac ns x))
  (exec-path-from-shell-initialize))


   #+END_SRC

   #+RESULTS:
   | bash | cannot set terminal process group (-1) | Inappropriate ioctl for devicebash | no job control in this shell[3;J[H[2J/home/cescoferraro/.nvm/versions/node/v6.10.2/bin | bash | cannot set terminal process group (-1) | Inappropriate ioctl for devicebash | no job control in this shell[3;J[H[2Jbash | cannot set terminal process group (-1) | Inappropriate ioctl for devicebash | no job control in this shell[3;J[H[2Jbash | cannot set terminal process group (-1) | Inappropriate ioctl for devicebash | no job control in this shell[3;J[H[2J/home/cescoferraro/.cargo/bin | /home/cescoferraro/bin | /usr/local/sbin | /usr/local/bin | /usr/sbin | /usr/bin | /sbin | /bin | /usr/games | /usr/local/games |

** Silence dumb warning
   #+BEGIN_SRC emacs-lisp
(setq ad-redefinition-action 'accept)
   #+END_SRC
** Force Fullscreen
   It forces emacs into full-screen-mode.
   #+BEGIN_SRC emacs-lisp
(add-hook 'window-setup-hook 'toggle-frame-fullscreen t)
   #+END_SRC

** Set Theme
   #+BEGIN_SRC emacs-lisp
     ;; (load-theme 'tsdh-dark)
(use-package color-theme
:ensure t)
;(use-package spacemacs-theme
;:ensure t
;:init
;(load-theme 'spacemacs-dark t))

(use-package zenburn-theme
:ensure t
:init (load-theme 'zenburn t) )

   #+END_SRC

** Hide ToolBar

   #+BEGIN_SRC emacs-lisp
(tool-bar-mode -1)
   #+END_SRC

** DEletion Mode
   #+BEGIN_SRC emacs-lisp
   (delete-selection-mode 1)
   #+END_SRC

   #+RESULTS:
   : t

** SimpleClipo
   #+BEGIN_SRC emacs-lisp
     (use-package simpleclip :ensure t
       :config

       (defun copy-to-x-clipboard ()
         (interactive)
         (let ((thing (if (region-active-p)
   		       (buffer-substring-no-properties (region-beginning) (region-end))
   		     (thing-at-point 'symbol))))
   	(simpleclip-set-contents thing)
   	(message "thing => clipboard!")))

       (defun paste-from-x-clipboard()
         "Paste string clipboard"
         (interactive)
         (insert (simpleclip-get-contents)))

       ;; Press `Alt-Y' to paste from clibpoard when in minibuffer
       (defun my/paste-in-minibuffer ()
         (local-set-key (kbd "M-y") 'paste-from-x-clipboard))
       (add-hook 'minibuffer-setup-hook 'my/paste-in-minibuffer)
       )
   #+END_SRC
** Autosave & backup files
   #+BEGIN_SRC emacs-lisp

(setq backup-directory-alist
      `((".*" . ,temporary-file-directory)))
(setq auto-save-file-name-transforms
      `((".*" ,temporary-file-directory t)))

   #+END_SRC
** Y&N

   #+BEGIN_SRC emacs-lisp
     (defalias 'yes-or-no-p 'y-or-n-p)
   #+END_SRC
** Narrow & WIdening

   #+BEGIN_SRC emacs-lisp
(defun narrow-or-widen-dwim (p)
  "If the buffer is narrowed, it widens. Otherwise, it narrows intelligently.
Intelligently means: region, org-src-block, org-subtree, or defun,
whichever applies first.
Narrowing to org-src-block actually calls `org-edit-src-code'.

With prefix P, don't widen, just narrow even if buffer is already
narrowed."
  (interactive "P")
  (declare (interactive-only))
  (cond ((and (buffer-narrowed-p) (not p)) (widen))
        ((region-active-p)
         (narrow-to-region (region-beginning) (region-end)))
        ((derived-mode-p 'org-mode)
         ;; `org-edit-src-code' is not a real narrowing command.
         ;; Remove this first conditional if you don't want it.
         (cond ((ignore-errors (org-edit-src-code))
                (delete-other-windows))
               ((org-at-block-p)
                (org-narrow-to-block))
               (t (org-narrow-to-subtree))))
        (t (narrow-to-defun))))

;; (define-key endless/toggle-map "n" #'narrow-or-widen-dwim)
;; This line actually replaces Emacs' entire narrowing keymap, that's
;; how much I like this command. Only copy it if that's what you want.
(define-key ctl-x-map "n" #'narrow-or-widen-dwim)

(eval-after-load 'org-src
  '(define-key org-src-mode-map
     "\C-x\C-s" #'org-edit-src-exit))

   #+END_SRC
** Prefer vertical window
   #+BEGIN_SRC emacs-lisp
(setq split-width-threshold 1 )
   #+END_SRC
* Org
  #+BEGIN_SRC emacs-lisp
(use-package org
  :ensure t
  :config

  (require 'ox-latex)
  (use-package org-bullets
    :ensure t
    :config


    (add-hook 'org-mode-hook (lambda () (org-bullets-mode t)))
    )

  (use-package ox-reveal
    :ensure ox-reveal)

  (setq org-reveal-root "http://cdn.jsdelivr.net/reveal.js/3.0.0/")
  (setq org-reveal-mathjax t)

  (use-package htmlize
    :ensure t)



  (add-hook 'org-mode-hook #'visual-line-mode)
  (diminish 'visual-line-mode)
  (setq org-src-fontify-natively t)
  ;; (setq org-support-shift-select t)
  (setq org-src-preserve-indentation t)
  (setq org-confirm-babel-evaluate nil)
  (org-babel-do-load-languages
   'org-babel-load-languages
   '((emacs-lisp . t)
     (python . t)
     (go . t)
     (java . t)
     (lisp . t)))

  )

  #+END_SRC
* Packages

** Move-text
   #+begin_src emacs-lisp
(use-package move-text
  :ensure t
  :config (progn
	    (global-set-key [(control shift j)] 'move-text-down)
	    (global-set-key [(control shift k)] 'move-text-up)
	    )
  )
   #+end_src

** Nyan Mode
   #+begin_src emacs-lisp
(use-package nyan-mode :ensure t :config (nyan-mode))
   #+end_src
** Emojis
   #+begin_src emacs-lisp
(use-package emojify
  :ensure t
  :config (progn
  (add-hook 'after-init-hook #'global-emojify-mode)))
(use-package emoji-cheat-sheet-plus
  :defer t
  :init
  (progn
    (global-set-key (kbd "C-c M-e") 'emoji-cheat-sheet-plus-insert)))
   #+end_src
** Helm

   #+BEGIN_SRC emacs-lisp
     (use-package helm
       :bind (
   	   ("M-x" . helm-M-x)
   	   ("C-x C-f" . helm-find-file)
   	   ))
(use-package helm-projectile
  :ensure t
  :config
  (global-set-key (kbd "C-x C-p") 'helm-projectile))
   #+END_SRC
** CSS

   #+BEGIN_SRC emacs-lisp
(use-package css-mode
  :mode (("\\.css\\'" . css-mode)
         ("\\.pcss\\'" . css-mode))
  :ensure t)

   #+END_SRC

** Rainbox
   #+BEGIN_SRC emacs-lisp
(use-package rainbow-delimiters
	  :ensure t)
(use-package rainbow-mode
  :ensure t
  :diminish rainbow-mode
  :config
  (add-hook 'prog-mode-hook 'rainbow-mode))
   #+END_SRC

** Openwith

   #+BEGIN_SRC emacs-lisp
(use-package openwith
	  :ensure t
	  :config
	  (openwith-mode t)
	  (setq openwith-associations '(("\\.pdf\\'" "evince" (file)))))
   #+END_SRC

** Iedit

   #+BEGIN_SRC emacs-lisp
(use-package iedit :ensure t)
   #+END_SRC
** Editorconfig
   #+BEGIN_SRC emacs-lisp
(use-package editorconfig
   :diminish editorconfig-mode
   :config
  (editorconfig-mode 1))
   #+END_SRC

** Projectile
   #+BEGIN_SRC emacs-lisp
(use-package projectile
	  :config
  (setq projectile-completion-system 'helm)
	  (projectile-global-mode))
   #+END_SRC

** Emmet
   #+BEGIN_SRC emacs-lisp
     (use-package emmet-mode
       :ensure t
       :diminish emmet-mode
       :init
       (add-hook 'html-mode-hook 'emmet-mode)
       (add-hook 'web-mode-hook 'emmet-mode))
   #+END_SRC
** Linum

   #+BEGIN_SRC emacs-lisp

     (use-package linum-relative
       :diminish linum-relative-mode
       :ensure t
       :config
       (progn
         (add-hook 'prog-mode-hook 'linum-mode)
         (add-hook 'prog-mode-hook 'linum-relative-global-mode)
         (setq linum-relative-current-symbol "")
         )
       )

   #+END_SRC
** Flycheck

   #+BEGIN_SRC emacs-lisp

(use-package flycheck
  :diminish flycheck-mode
  :bind (("M-n" . flycheck-next-error) ("M-p" . flycheck-previous-error))
  :config
  (global-flycheck-mode)
  (flycheck-add-mode 'typescript-tslint 'web-mode)
  (setq-default flycheck-disabled-checkers '(emacs-lisp-checkdoc))
  (setq flycheck-display-errors-delay 0)
  (setq flycheck-shellcheck-follow-sources nil)

  (set-face-attribute 'flycheck-error nil :background "#8c5353")
  (set-face-attribute 'flycheck-warning nil :background "#535399")
  )
   #+END_SRC
** GIT

   #+BEGIN_SRC emacs-lisp
(use-package git-gutter-fringe
  :diminish git-gutter-mode
  :config
  (global-git-gutter-mode t)
  (set-face-foreground 'git-gutter-fr:modified "yellow")
  (set-face-foreground 'git-gutter-fr:added    "blue")
  (set-face-foreground 'git-gutter-fr:deleted  "white")
  (setq git-gutter-fr:side 'left-fringe))


(defun cesco/refresh ()
  (interactive)
  (kill-this-buffer)
  (magit-list-repositories)
  )

(defun cesco/magit-repolist-delete (&optional _button)
  "Show the status for the repository at point."
  (interactive)
  (--if-let (tabulated-list-get-id)
      (if (yes-or-no-p "Remove Repo? ")
	  (progn
	    (delete-directory (expand-file-name it) t t)
	    (cesco/refresh)
	    )
	)
    (user-error "There is no repository at point")))

(defun cesco/magit-repolist-hook ()
  (local-set-key (kbd "C-M-g")  'magit-repolist-status)
  (evil-define-key
    (list 'normal 'insert 'visual 'motion)
    magit-repolist-mode-map
    (kbd "TAB") 'magit-repolist-status
    (kbd "q") 'kill-this-buffer
    (kbd "d") 'cesco/magit-repolist-delete
    (kbd "r") 'cesco/refresh
    (kbd "RET") 'magit-repolist-status)
  (message "cesco/magit-repolist-hook")
  )

(use-package magit
  :after evil-leader
  :bind ("C-c r" . cesco/automatic-push)
  :config
  (setq magit-auto-revert-mode nil)
  (setq magit-repository-directories
	`(("~/go/src/github.com/cescoferraro" . 1)
	  (,user-emacs-directory              . 1)))

  (setq magit-repolist-columns
	'(("Name"    25 magit-repolist-column-ident                  ())
	  ("D"        1 magit-repolist-column-dirty                  ())
	  ("Behind "    5  magit-repolist-column-unpulled-from-upstream ((:right-align t)))
	  ("Ahead "    5  magit-repolist-column-unpushed-to-upstream   ((:right-align t)))
	  ("Version" 25 magit-repolist-column-version                ())
	  ))

  ;; Evil Keybinding
  (evil-leader/set-key "g" 'magit-status)
  (add-hook 'magit-repolist-mode-hook 'cesco/magit-repolist-hook)
  )

(use-package magit-gitflow
  :after (magit)
  :config
  (evil-define-key
    'normal magit-mode-map (kbd ">") 'magit-gitflow-popup)
  :init
  (add-hook 'magit-mode-hook 'turn-on-magit-gitflow))

(use-package evil-magit
  :after magit
  :ensure t )

   #+END_SRC

** Hl-line+
   #+BEGIN_SRC emacs-lisp
   #+END_SRC
** Which-Keys
   #+BEGIN_SRC emacs-lisp
(use-package which-key :ensure t :diminish which-key-mode :config (which-key-mode))
   #+END_SRC
** Twitter

   #+BEGIN_SRC emacs-lisp
(use-package twittering-mode
  :config
  (setq twittering-use-master-password t)
  (setq twittering-icon-mode t)
  (setq twittering-timer-interval 300)
  (setq twittering-url-show-status nil))
   #+END_SRC

** Neotree
   #+BEGIN_SRC emacs-lisp
(defun cesco/extension-icons ()
  (add-to-list 'all-the-icons-icon-alist
	       '("^dobi.yaml"
		 all-the-icons-alltheicon "terminal"
		 :height 1.0
		 :face all-the-icons-pink))
  (add-to-list 'all-the-icons-icon-alist
	       '("\\.pcss"
		 all-the-icons-alltheicon "css3"
		 :height 1.0
		 :face all-the-icons-red))
  (add-to-list 'all-the-icons-icon-alist
	       '("\\.tsx$"
		 all-the-icons-alltheicon "react"
		 :height 1.0
		 :face all-the-icons-blue))
  )

(defun cesco/neotree-copy ()
  (neotree-copy-filepath-to-yank-ring)
  (message "Path copied to the kill ring!")
  )

(defun cesco/neotree-toogle ()
  (interactive)
  (if (string-equal " *NeoTree*" (buffer-name))
      (progn
	(delete-window)
	(kill-buffer " *NeoTree*")
	(message "have just deleted neotree window"))
    (progn
      (if (get-buffer " *NeoTree*")
	  (progn
	    (message "neotree buffer was opened")
	    (kill-buffer " *NeoTree*"))
	(progn
	  (if (projectile-project-p)
	      (neotree-projectile-action)
	    (neotree-toggle))
	  (message "have just popped neotree")))))
  )

(defun cesco/neotree-evil ()
  (define-key evil-normal-state-local-map (kbd "q") 'neotree-hide)
  (define-key evil-normal-state-local-map (kbd "I") 'neotree-hidden-file-toggle)
  (define-key evil-normal-state-local-map (kbd "z") 'neotree-stretch-toggle)
  (define-key evil-normal-state-local-map (kbd "1") 'neotree-change-root)
  (define-key evil-normal-state-local-map (kbd "R") 'neotree-refresh)
  (define-key evil-normal-state-local-map (kbd "m") 'neotree-rename-node)
  (define-key evil-normal-state-local-map (kbd "n") 'neotree-create-node)
  (define-key evil-normal-state-local-map (kbd "c") 'neotree-copy-node)
  (define-key evil-normal-state-local-map (kbd "C") 'cesco/neotree-copy)
  (define-key evil-normal-state-local-map (kbd "d") 'neotree-delete-node)
  (define-key evil-normal-state-local-map (kbd "s") 'neotree-enter-vertical-split)
  (define-key evil-normal-state-local-map (kbd "S") 'neotree-enter-horizontal-split)
  (define-key evil-normal-state-local-map (kbd "RET") 'neotree-enter)
  (define-key evil-normal-state-local-map (kbd "TAB") 'neotree-enter))

(use-package find-file-in-project :ensure t)
(use-package all-the-icons :ensure t :config (cesco/extension-icons))
(use-package neotree
  :after (evil evil-leader)
  :ensure all-the-icons
  :bind ([f2] . neotree-projectile-action)
  :config (evil-leader/set-key "m" 'cesco/neotree-toogle)
  :init (progn
	  (setq neo-theme (if window-system 'icons 'nerd)) ; 'classic, 'nerd, 'ascii, 'arrow
	  (setq neo-show-hidden-files t)
	  (add-hook 'neotree-mode-hook 'cesco/neotree-evil)))
   #+END_SRC

   #+RESULTS:

** Ivy
   #+BEGIN_SRC emacs-lisp
     (use-package counsel :ensure t)
     (use-package swiper
       :diminish ivy-mode
       :ensure t
       :config
       (progn
         (ivy-mode 1)
         (setq ivy-use-virtual-buffers t)
         (setq enable-recursive-minibuffers t)
         (global-set-key "\C-s" 'swiper)
         (global-set-key (kbd "C-c C-r") 'ivy-resume)
         (global-set-key (kbd "<f6>") 'ivy-resume)
         (global-set-key (kbd "<f1> f") 'counsel-describe-function)
         (global-set-key (kbd "<f1> v") 'counsel-describe-variable)
         (global-set-key (kbd "<f1> l") 'counsel-find-library)
         ;; (global-set-key (kbd "<f2> i") 'counsel-info-lookup-symbol)
         ;; (global-set-key (kbd "<f2> u") 'counsel-unicode-char)
         (global-set-key (kbd "C-c k") 'counsel-ag)
         (global-set-key (kbd "C-x l") 'counsel-locate)
         (global-set-key (kbd "C-S-o") 'counsel-rhythmbox)
         (define-key read-expression-map (kbd "C-r") 'counsel-expression-history)
         )
       )


   #+END_SRC
** Golden-ratio
   #+BEGIN_SRC emacs-lisp
     (use-package golden-ratio
       :diminish golden-ratio-mode
   	    :config
   	    (golden-ratio-mode 1))
   #+END_SRC
** Beacon
   #+BEGIN_SRC emacs-lisp
(use-package beacon
:diminish beacon-mode
:ensure t
:config
(progn
(beacon-mode 1)
(setq beacon-push-mark 35)
(setq beacon-color "#666600")))
   #+END_SRC
** Eldoc

   #+begin_src emacs-lisp
(use-package eldoc :ensure t :diminish eldoc-mode)
   #+end_src


** ModeLine
   #+BEGIN_SRC emacs-lisp
(use-package spaceline :ensure t
  :config
  (setq-default
   powerline-height 24
   powerline-default-separator 'arrow
   spaceline-flycheck-bullet "❖ %s"
   spaceline-separator-dir-left '(left . left)
   spaceline-separator-dir-right '(right . right))
  (set-face-attribute 'spaceline-evil-emacs nil :background "#be84ff"  )
  (set-face-attribute 'spaceline-evil-insert nil :background "#5fd7ff" :foreground "black")
  (set-face-attribute 'spaceline-evil-motion nil :background "#ae81ff ")
  (set-face-attribute 'spaceline-evil-normal nil :background "#FF5722" :foreground "white")
  (set-face-attribute 'spaceline-evil-replace nil :background "#f92672")
  (set-face-attribute 'spaceline-evil-visual nil :background "#fd971f")
  (set-face-attribute 'mode-line-buffer-id nil :foreground "#E1BEE7" :background "black")
  (set-face-attribute 'mode-line-emphasis  nil :foreground "white" :background "#7B1FA2")
  (setq-default mode-line-format '("%e" (:eval (spaceline-ml-main))))
  (setq-default spaceline-highlight-face-func 'spaceline-highlight-face-evil-state)
  )

(use-package spaceline-config
  :ensure spaceline
  :config
  (spaceline-install
   'main
   '(

src/emacs.d/configuration.el
(setq user-full-name "Francesco Antonello Ferraro"
      user-mail-address "francescoaferraro@gmail.com "
      calendar-latitude 42.2
      calendar-longitude -71.1
      calendar-location-name "Cambridge, MA")

(load-theme 'tsdh-dark)

(tool-bar-mode -1)

(use-package editorconfig
  :config
  (editorconfig-mode 1))

(use-package projectile
	  :config
	  (projectile-global-mode))

(use-package linum-relative
  :diminish linum-relative-mode
  :ensure t
  :config
  (progn (linum-mode)
         (linum-relative-global-mode)
         (setq linum-relative-current-symbol "")
         )
)

(use-package spaceline
    :config
    (require 'spaceline-config)
    (spaceline-spacemacs-theme)
    (spaceline-toggle-buffer-size-off)
    (spaceline-toggle-version-control-on)
    (spaceline-toggle-selection-info-on)
    (spaceline-toggle-buffer-encoding-on)
    (spaceline-toggle-minor-modes-on)
    (spaceline-toggle-minor-modes-on))

(use-package flycheck
  :init
  ;(add-hook 'after-init-hook #'global-flycheck-mode)
  :config
  (setq flycheck-display-errors-delay 0.3)
  (set-face-attribute 'flycheck-error nil :background "#8c5353")
  (set-face-attribute 'flycheck-warning nil :background "#535399")
  :bind (("M-n" . flycheck-next-error)
         ("M-p" . flycheck-previous-error))
)

(use-package magit
  :ensure magit
  :ensure evil-magit
  :after evil
  :init
  (evil-magit-init)
  (evil-leader/set-key "g" 'magit-status))

(add-to-list 'load-path (concat (getenv "GOPATH")  "/src/github.com/golang/lint/misc/emacs"))
(require 'golint)
(require 'godoctor)
;; golang shorcuts
(require 'evil-leader)
(evil-leader/set-key-for-mode 'go-mode
  "j" 'godef-jump-other-window
  "g" 'godef-jump
 )





(defun set-exec-path-from-shell-PATH ()
  (let ((path-from-shell (replace-regexp-in-string
                          "[ \t\n]*$"
                          ""
                          (shell-command-to-string "$SHELL --login -i -c 'echo $PATH'"))))
    (setenv "PATH" path-from-shell)
    (setq eshell-path-env path-from-shell) ; for eshell users
    (setq exec-path (split-string path-from-shell path-separator))))

(when window-system (set-exec-path-from-shell-PATH))

(setenv "GOPATH" "/home/cescoferraro/go")
(add-to-list 'exec-path "/home/cescoferraro/go/bin")


(defun my-go-mode-hook ()
  (setq gofmt-command "goimports")
  ; Call Gofmt before saving
  (add-hook 'before-save-hook 'gofmt-before-save)
  ; Customize compile command to run go build
  (if (not (string-match "go" compile-command))
      (set (make-local-variable 'compile-command)
           "go build -v && go test -v && go vet"))
  (load-file "$GOPATH/src/github.com/dominikh/go-mode.el/go-guru.el")
  )

(defun auto-complete-for-go ()
  (auto-complete-mode 1))

(with-eval-after-load 'go-mode
  (lambda ()
   (require 'go-autocomplete))
)



(add-hook 'go-mode-hook 'go-eldoc-setup)
(add-hook 'go-mode-hook 'auto-complete-for-go)
(add-hook 'go-mode-hook 'my-go-mode-hook)

(require 'meghanada)
(require 'flycheck)
(require 'evil-leader)

;;; Code:

(evil-leader/set-key-for-mode 'java-mode "t"
    (lambda ()
        (interactive)
        (meghanada-run-task "test")
        (ace-window "")
        (golden-ratio)
        )
    )
(evil-leader/set-key-for-mode 'java-mode "c"
    (lambda ()
        (interactive)
        (meghanada-run-task "run")
        (ace-window "")
        (golden-ratio)
        )
    )


(add-hook 'java-mode-hook
    (lambda ()
        ;; meghanada-mode on
        (meghanada-mode t)
        (add-hook 'before-save-hook (lambda ()
                                        (flycheck-display-error-at-point)
                                        (meghanada-code-beautify-before-save)
                                        ))))

(require 'evil-leader)

;; typescript shortcuts
(evil-leader/set-key-for-mode 'web-mode
  "j"  (lambda ()
    (interactive)
    (tide-jump-to-implementation )
    )
  "g"  (lambda ()
    (interactive)
    (tide-jump-to-definition )
    )
)


(defun setup-tide-mode ()
  (interactive)
  (tide-setup)
  (flycheck-mode +1)
  (setq flycheck-check-syntax-automatically '(save mode-enabled))
  (eldoc-mode +1)
  (tide-hl-identifier-mode +1)
  ;; company is an optional dependency. You have to
  ;; install it separately via package-install
  ;; `M-x package-install [ret] company`
  (company-mode +1))

;; aligns annotation to the right hand side
(setq company-tooltip-align-annotations t)

;; formats the buffer before saving
(add-hook 'before-save-hook 'tide-format-before-save)

(add-hook 'typescript-mode-hook #'setup-tide-mode)


(require 'web-mode)
(add-to-list 'auto-mode-alist '("\\.tsx\\'" . web-mode))
(add-hook 'web-mode-hook
	  (lambda ()
	    (when (string-equal "tsx" (file-name-extension buffer-file-name))
	                    (setup-tide-mode))))

(add-to-list 'auto-mode-alist '("\\.jsx\\'" . web-mode))
(add-hook 'web-mode-hook
    (lambda ()
        (when (string-equal "jsx" (file-name-extension buffer-file-name))
                          (setup-tide-mode))))

(use-package evil
  :ensure t
  :config
  (use-package evil-org :ensure t)
  (use-package evil-surround
    :ensure t
    :config
    (global-evil-surround-mode))
  (use-package evil-leader
    :ensure t
    :config
    (evil-leader/set-leader ",")
    (evil-leader/set-key
    "m"  'neotree-toggle
    "n"  'neotree-project-dir
     "v" 'evil-visual-block
     "z" 'zoom-window-zoom
     "f" 'helm-projectile
     "s" 'save-buffer
     "q" 'kill-emacs
     "G" 'magit-status
     "w" 'delete-window
     "e" 'kill-this-buffer
     "E" 'org-export-dispatch
     "," 'helm-projectile-switch-to-buffer
     "'"  (lambda () (interactive) (ansi-term (getenv "SHELL")))
     (kbd ".") (lambda () (interactive (ace-window "")))
     (kbd "o") (lambda () (interactive) (find-file "~/.bash.local"))
     (kbd "i") (lambda () (interactive) (find-file "~/.emacs.d/init.el"))
     (kbd "/") ( lambda () (interactive)
          (comment-or-uncomment-region (line-beginning-position) (line-end-position)))
    )

    (evil-define-key 'visual evil-surround-mode-map "A" 'evil-substitute)
    (evil-define-key 'visual evil-surround-mode-map "a" 'evil-surround-region)
    (evil-define-key 'visual evil-surround-mode-map "E" 'org-export-dispatch)
    (define-key evil-normal-state-map (kbd "TAB") 'org-cycle)
    )

(use-package neotree
  :init
(add-hook 'neotree-mode-hook
    (lambda ()
      (define-key evil-normal-state-local-map (kbd "q") 'neotree-hide)
      (define-key evil-normal-state-local-map (kbd "I") 'neotree-hidden-file-toggle)
      (define-key evil-normal-state-local-map (kbd "z") 'neotree-stretch-toggle)
    (define-key evil-normal-state-local-map (kbd "1") 'neotree-change-root)
      (define-key evil-normal-state-local-map (kbd "R") 'neotree-refresh)
      (define-key evil-normal-state-local-map (kbd "m") 'neotree-rename-node)
      (define-key evil-normal-state-local-map (kbd "c") 'neotree-create-node)
      (define-key evil-normal-state-local-map (kbd "d") 'neotree-delete-node)

      (define-key evil-normal-state-local-map (kbd "s") 'neotree-enter-vertical-split)
      (define-key evil-normal-state-local-map (kbd "S") 'neotree-enter-horizontal-split)

      (define-key evil-normal-state-local-map (kbd "RET") 'neotree-enter)
      (define-key evil-normal-state-local-map (kbd "TAB") 'neotree-enter)
))
)

  :config
  (global-evil-leader-mode)
  (evil-mode)

  ; Search by symbols rather than words by default
  (set-default 'evil-symbol-word-search t)
  )



(use-package counsel :ensure t)
(use-package swiper
    :ensure t
    :config
    (progn
        (ivy-mode 1)
        (setq ivy-use-virtual-buffers t)
        (setq enable-recursive-minibuffers t)
        (global-set-key "\C-s" 'swiper)
        (global-set-key (kbd "C-c C-r") 'ivy-resume)
        (global-set-key (kbd "<f6>") 'ivy-resume)
        (global-set-key (kbd "M-x") 'counsel-M-x)
        (global-set-key (kbd "C-x C-f") 'counsel-find-file)
        (global-set-key (kbd "<f1> f") 'counsel-describe-function)
        (global-set-key (kbd "<f1> v") 'counsel-describe-variable)
        (global-set-key (kbd "<f1> l") 'counsel-find-library)
        (global-set-key (kbd "<f2> i") 'counsel-info-lookup-symbol)
        (global-set-key (kbd "<f2> u") 'counsel-unicode-char)
        (global-set-key (kbd "C-c g") 'counsel-git)
        (global-set-key (kbd "C-c j") 'counsel-git-grep)
        (global-set-key (kbd "C-c k") 'counsel-ag)
        (global-set-key (kbd "C-x l") 'counsel-locate)
        (global-set-key (kbd "C-S-o") 'counsel-rhythmbox)
        (define-key read-expression-map (kbd "C-r") 'counsel-expression-history)
    )
)
